[Bug] issue installing `luasystem`

```
lua: ...9a0a7b27412c333e7d-luasystem@0.4.4-1/src/system/init.lua:7: module 'system.core' not found:
        no field package.preload['system.core']
...
```

This appears to be an issue with the builtin build.

Found the cause:

```
/nix/store/pbixrqd120nnw8a0ygfpm3mnwlr1mbj7-binutils-2.43.1/bin/ld: cannot find -l/tmp/luasystem.ocwcZ202bWXP/rt: No such file or directory
```

--- a/rocks-lib/src/build/builtin.rs
+++ b/rocks-lib/src/build/builtin.rs
@@ -6,7 +6,7 @@ use std::{
 use crate::{
     config::Config,
     lua_installation::LuaInstallation,
-    rockspec::{utils, Build, BuiltinBuildSpec, ModuleSpec},
+    rockspec::{utils, Build, BuiltinBuildSpec, LuaModule, ModuleSpec},
     tree::RockLayout,
 };
 use indicatif::{MultiProgress, ProgressBar};
@@ -92,7 +92,7 @@ impl Build for BuiltinBuildSpec {
     }
 }
 
-fn autodetect_modules(build_dir: &Path) -> HashMap<String, ModuleSpec> {
+fn autodetect_modules(build_dir: &Path) -> HashMap<LuaModule, ModuleSpec> {
     WalkDir::new(build_dir.join("src"))
         .into_iter()
         .chain(WalkDir::new(build_dir.join("lua")))
@@ -124,15 +124,10 @@ fn autodetect_modules(build_dir: &Path) -> HashMap<String, ModuleSpec> {
             // The rockspec requires the format to be like this, and representing our
             // data in this form allows us to respect any overrides made by the user (which follow
             // the `module.name` format, not our internal one).
-            let lua_module_path = diff
-                .components()
-                .skip(1)
-                .collect::<PathBuf>()
-                .to_string_lossy()
-                .trim_end_matches(".lua")
-                .replace(std::path::MAIN_SEPARATOR_STR, ".");
+            let pathbuf = diff.components().skip(1).collect::<PathBuf>();
+            let lua_module = LuaModule::from_pathbuf(pathbuf);
 
-            (lua_module_path, ModuleSpec::SourcePath(diff))
+            (lua_module, ModuleSpec::SourcePath(diff))
         })
         .collect()
 }

--- a/rocks-lib/src/build/mod.rs
+++ b/rocks-lib/src/build/mod.rs
@@ -63,7 +63,7 @@ async fn run_build(
     build_dir: &Path,
 ) -> Result<(), BuildError> {
     with_spinner(progress, "🛠️ Building...".into(), || async {
-        match rockspec.build.default.build_backend.to_owned() {
+        match rockspec.build.current_platform().build_backend.to_owned() {
             Some(BuildBackendSpec::Builtin(build_spec)) => {
                 build_spec
                     .run(progress, output_paths, false, lua, config, build_dir)

--- a/rocks-lib/src/rockspec/build/builtin.rs
+++ b/rocks-lib/src/rockspec/build/builtin.rs
@@ -1,6 +1,6 @@
 use itertools::Itertools as _;
 use serde::{de, Deserialize, Deserializer};
-use std::{collections::HashMap, convert::Infallible, path::PathBuf};
+use std::{collections::HashMap, convert::Infallible, fmt::Display, path::PathBuf, str::FromStr};
 use thiserror::Error;
 
 use crate::rockspec::{
@@ -11,7 +11,59 @@ use crate::rockspec::{
 #[derive(Debug, PartialEq, Deserialize, Default, Clone)]
 pub struct BuiltinBuildSpec {
     /// Keys are module names in the format normally used by the `require()` function
-    pub modules: HashMap<String, ModuleSpec>,
+    pub modules: HashMap<LuaModule, ModuleSpec>,
+}
+
+#[derive(Debug, PartialEq, Eq, Deserialize, Default, Clone, Hash)]
+pub struct LuaModule(String);
+
+impl LuaModule {
+    pub fn to_lua_path(&self) -> PathBuf {
+        self.to_pathbuf(".lua")
+    }
+
+    pub fn to_lib_path(&self) -> PathBuf {
+        self.to_pathbuf(std::env::consts::DLL_SUFFIX)
+    }
+
+    fn to_pathbuf(&self, extension: &str) -> PathBuf {
+        PathBuf::from(self.0.replace('.', std::path::MAIN_SEPARATOR_STR) + extension)
+    }
+
+    pub fn from_pathbuf(path: PathBuf) -> Self {
+        let extension = path
+            .extension()
+            .map(|ext| ext.to_string_lossy().to_string())
+            .unwrap_or("".into());
+        let module = path
+            .to_string_lossy()
+            .trim_end_matches(extension.as_str())
+            .replace(std::path::MAIN_SEPARATOR_STR, ".");
+        LuaModule(module)
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+}
+
+#[derive(Error, Debug)]
+#[error("could not parse lua module from {0}.")]
+pub struct ParseLuaModuleError(String);
+
+impl FromStr for LuaModule {
+    type Err = ParseLuaModuleError;
+
+    // NOTE: We may want to add some validations
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(LuaModule(s.into()))
+    }
+}
+
+impl Display for LuaModule {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
 }
 
 #[derive(Debug, PartialEq, Clone)]
@@ -260,14 +312,23 @@ mod tests {
         lua.load(lua_content).exec().unwrap();
         let build_spec: BuiltinBuildSpec =
             lua.from_value(lua.globals().get("build").unwrap()).unwrap();
-        let foo = build_spec.modules.get("foo").unwrap();
+        let foo = build_spec
+            .modules
+            .get(&LuaModule::from_str("foo").unwrap())
+            .unwrap();
         assert_eq!(*foo, ModuleSpec::SourcePath("lua/foo/init.lua".into()));
-        let bar = build_spec.modules.get("bar").unwrap();
+        let bar = build_spec
+            .modules
+            .get(&LuaModule::from_str("bar").unwrap())
+            .unwrap();
         assert_eq!(
             *bar,
             ModuleSpec::SourcePaths(vec!["lua/bar.lua".into(), "lua/bar/internal.lua".into()])
         );
-        let baz = build_spec.modules.get("baz").unwrap();
+        let baz = build_spec
+            .modules
+            .get(&LuaModule::from_str("baz").unwrap())
+            .unwrap();
         assert!(matches!(baz, ModuleSpec::ModulePaths { .. }));
         let lua_content_no_sources = "
         build = {\n
@@ -297,7 +358,10 @@ mod tests {
         lua.load(lua_content_complex_defines).exec().unwrap();
         let build_spec: BuiltinBuildSpec =
             lua.from_value(lua.globals().get("build").unwrap()).unwrap();
-        let baz = build_spec.modules.get("baz").unwrap();
+        let baz = build_spec
+            .modules
+            .get(&LuaModule::from_str("baz").unwrap())
+            .unwrap();
         match baz {
             ModuleSpec::ModulePaths(paths) => assert_eq!(
                 paths.defines,

--- a/rocks-lib/src/rockspec/build/mod.rs
+++ b/rocks-lib/src/rockspec/build/mod.rs
@@ -4,13 +4,16 @@ mod make;
 mod rust_mlua;
 
 pub mod utils; // Make build utilities available as a submodule
-pub use builtin::{BuiltinBuildSpec, ModulePaths, ModuleSpec};
+pub use builtin::{BuiltinBuildSpec, LuaModule, ModulePaths, ModuleSpec};
 pub use cmake::*;
 use indicatif::MultiProgress;
 pub use make::*;
 pub use rust_mlua::*;
 
-use builtin::{ModulePathsMissingSources, ModuleSpecAmbiguousPlatformOverride, ModuleSpecInternal};
+use builtin::{
+    ModulePathsMissingSources, ModuleSpecAmbiguousPlatformOverride, ModuleSpecInternal,
+    ParseLuaModuleError,
+};
 
 use itertools::Itertools as _;
 
@@ -20,6 +23,7 @@ use std::{
     env::consts::DLL_EXTENSION,
     future::Future,
     path::{Path, PathBuf},
+    str::FromStr,
 };
 use thiserror::Error;
 
@@ -63,6 +67,8 @@ pub enum BuildSpecInternalError {
     InvalidRustMLuaFormat,
     #[error(transparent)]
     ModulePathsMissingSources(#[from] ModulePathsMissingSources),
+    #[error(transparent)]
+    ParseLuaModuleError(#[from] ParseLuaModuleError),
 }
 
 impl BuildSpec {
@@ -78,14 +84,14 @@ impl BuildSpec {
                             LuaTableKey::IntKey(_) => {
                                 Err(BuildSpecInternalError::ModulesHaveListElements)
                             }
-                            LuaTableKey::StringKey(str) => Ok(str),
+                            LuaTableKey::StringKey(str) => Ok(LuaModule::from_str(str.as_str())?),
                         }?;
                         match ModuleSpec::from_internal(module_spec_internal) {
                             Ok(module_spec) => Ok((key_str, module_spec)),
                             Err(err) => Err(err.into()),
                         }
                     })
-                    .collect::<Result<HashMap<String, ModuleSpec>, BuildSpecInternalError>>()?,
+                    .collect::<Result<HashMap<LuaModule, ModuleSpec>, BuildSpecInternalError>>()?,
             })),
             BuildType::Make => {
                 let default = MakeBuildSpec::default();
@@ -223,10 +229,10 @@ pub struct CommandBuildSpec {
 pub struct InstallSpec {
     /// Lua modules written in Lua.
     #[serde(default)]
-    pub lua: HashMap<String, PathBuf>,
+    pub lua: HashMap<LuaModule, PathBuf>,
     /// Dynamic libraries implemented compiled Lua modules.
     #[serde(default)]
-    pub lib: HashMap<String, PathBuf>,
+    pub lib: HashMap<LuaModule, PathBuf>,
     /// Configuration files.
     #[serde(default)]
     pub conf: HashMap<String, PathBuf>,

--- a/rocks-lib/src/rockspec/build/utils/mod.rs
+++ b/rocks-lib/src/rockspec/build/utils/mod.rs
@@ -7,21 +7,16 @@ use target_lexicon::Triple;
 
 use crate::{build::BuildError, lua_installation::LuaInstallation};
 
-use super::ModulePaths;
-
-fn lua_module_to_pathbuf(path: &str, extension: &str) -> PathBuf {
-    PathBuf::from(path.replace('.', std::path::MAIN_SEPARATOR_STR) + extension)
-}
+use super::{LuaModule, ModulePaths};
 
 /// Copies a lua source file to a specific destination. The destination is described by a
 /// `module.path` syntax (equivalent to the syntax provided to Lua's `require()` function).
 pub fn copy_lua_to_module_path(
     source: &PathBuf,
-    target_module_name: &str,
+    target_module: &LuaModule,
     target_dir: &Path,
 ) -> io::Result<()> {
-    let target = lua_module_to_pathbuf(target_module_name, ".lua");
-    let target = target_dir.join(target);
+    let target = target_dir.join(target_module.to_lua_path());
 
     std::fs::create_dir_all(target.parent().unwrap())?;
 
@@ -35,12 +30,11 @@ pub fn copy_lua_to_module_path(
 /// Panics if no parent or no filename can be determined for the target path.
 pub fn compile_c_files(
     files: &Vec<PathBuf>,
-    target_file: &str,
+    target_module: &LuaModule,
     target_dir: &Path,
     lua: &LuaInstallation,
 ) -> Result<(), BuildError> {
-    let target = lua_module_to_pathbuf(target_file, std::env::consts::DLL_SUFFIX);
-    let target = target_dir.join(target);
+    let target = target_dir.join(target_module.to_lib_path());
 
     let parent = target.parent().unwrap_or_else(|| {
         panic!(
@@ -62,6 +56,7 @@ pub fn compile_c_files(
     // See https://github.com/rust-lang/cc-rs/issues/594#issuecomment-2110551057
 
     let mut build = cc::Build::new();
+    let intermediate_dir = tempdir::TempDir::new(target_module.as_str())?;
     let build = build
         .cargo_metadata(false)
         .debug(false)
@@ -69,7 +64,7 @@ pub fn compile_c_files(
         .host(std::env::consts::OS)
         .includes(&lua.include_dir)
         .opt_level(3)
-        .out_dir(parent)
+        .out_dir(intermediate_dir)
         .target(&host.to_string());
     let objects = build.compile_intermediates();
     build
@@ -97,14 +92,19 @@ pub fn lua_lib_extension() -> &'static str {
 pub fn compile_c_modules(
     data: &ModulePaths,
     source_dir: &Path,
-    target_file: &str,
+    target_module: &LuaModule,
     target_dir: &Path,
     lua: &LuaInstallation,
 ) -> Result<(), BuildError> {
-    let target = lua_module_to_pathbuf(target_file, std::env::consts::DLL_SUFFIX);
-    let target = target_dir.join(target);
+    let target = target_dir.join(target_module.to_lib_path());
 
-    std::fs::create_dir_all(target.parent().unwrap())?;
+    let parent = target.parent().unwrap_or_else(|| {
+        panic!(
+            "Couldn't determine parent for path {}",
+            target.to_str().unwrap_or("")
+        )
+    });
+    std::fs::create_dir_all(parent)?;
 
     let host = Triple::host();
 
@@ -119,6 +119,8 @@ pub fn compile_c_modules(
         .iter()
         .map(|dir| source_dir.join(dir))
         .collect_vec();
+
+    let intermediate_dir = tempdir::TempDir::new(target_module.as_str())?;
     let build = build
         .cargo_metadata(false)
         .debug(false)
@@ -127,7 +129,7 @@ pub fn compile_c_modules(
         .includes(&include_dirs)
         .includes(&lua.include_dir)
         .opt_level(3)
-        .out_dir(target_dir)
+        .out_dir(intermediate_dir)
         .shared_flag(true)
         .target(&host.to_string());
 
@@ -137,22 +139,34 @@ pub fn compile_c_modules(
         build.define(name, value.as_deref());
     }
 
-    for libdir in &data.libdirs {
-        build.flag(&("-L".to_string() + source_dir.join(libdir).to_str().unwrap()));
-    }
-
-    for library in &data.libraries {
-        build.flag(&("-l".to_string() + source_dir.join(library).to_str().unwrap()));
-    }
-
     let file = target.file_name().unwrap_or_else(|| {
         panic!(
             "Couldn't determine filename for path {}",
             target.to_str().unwrap_or("")
         )
     });
+    // See https://github.com/rust-lang/cc-rs/issues/594#issuecomment-2110551057
+    let objects = build.compile_intermediates();
 
-    build.try_compile(file.to_str().unwrap())?;
+    let libdir_args = data
+        .libdirs
+        .iter()
+        .map(|libdir| format!("-L{}", source_dir.join(libdir).to_str().unwrap()));
+
+    let library_args = data
+        .libraries
+        .iter()
+        .map(|library| format!("-l{}", library.to_str().unwrap()));
+
+    build
+        .get_compiler()
+        .to_command()
+        .args(["-shared", "-o"])
+        .arg(parent.join(file))
+        .args(&objects)
+        .args(libdir_args)
+        .args(library_args)
+        .status()?;
 
     Ok(())
 }

--- a/rocks-lib/src/rockspec/mod.rs
+++ b/rocks-lib/src/rockspec/mod.rs
@@ -533,7 +533,13 @@ mod tests {
             rockspec.build.default.build_backend,
             Some(BuildBackendSpec::Builtin { .. })
         ));
-        let foo_bar_path = rockspec.build.default.install.lua.get("foo.bar").unwrap();
+        let foo_bar_path = rockspec
+            .build
+            .default
+            .install
+            .lua
+            .get(&LuaModule::from_str("foo.bar").unwrap())
+            .unwrap();
         assert_eq!(*foo_bar_path, PathBuf::from("src/bar.lua"));
         let foo_bar_path = rockspec.build.default.install.bin.get("foo.bar").unwrap();
         assert_eq!(*foo_bar_path, PathBuf::from("bin/bar"));
@@ -608,7 +614,13 @@ mod tests {
             rockspec.build.default.build_backend,
             Some(BuildBackendSpec::Make(MakeBuildSpec::default()))
         );
-        let foo_bar_path = rockspec.build.default.install.lib.get("foo.bar").unwrap();
+        let foo_bar_path = rockspec
+            .build
+            .default
+            .install
+            .lib
+            .get(&LuaModule::from_str("foo.bar").unwrap())
+            .unwrap();
         assert_eq!(*foo_bar_path, PathBuf::from("lib/bar.so"));
         let copy_directories = rockspec.build.default.copy_directories;
         assert_eq!(
@@ -924,7 +936,7 @@ mod tests {
             win32.build_backend,
             Some(BuildBackendSpec::Builtin(BuiltinBuildSpec {
                 modules: vec![(
-                    "cjson".into(),
+                    LuaModule::from_str("cjson").unwrap(),
                     ModuleSpec::ModulePaths(ModulePaths {
                         sources: vec!["lua_cjson.c".into(), "strbuf.c".into(), "fpconv.c".into()],
                         libraries: Vec::default(),
@@ -1058,11 +1070,11 @@ build = {
             &build_spec.build_backend
         {
             assert_eq!(
-                modules.get("system.init"),
+                modules.get(&LuaModule::from_str("system.init").unwrap()),
                 Some(&ModuleSpec::SourcePath("system/init.lua".into()))
             );
             assert_eq!(
-                modules.get("system.core"),
+                modules.get(&LuaModule::from_str("system.core").unwrap()),
                 Some(&ModuleSpec::ModulePaths(ModulePaths {
                     sources: vec![
                         "src/core.c".into(),

